commit f7fd821f7046b055643cfb4a14b6585a2e3d989f
Author: chiachin <chiachin@netivism.com.tw>
Date:   Wed Apr 2 17:12:17 2025 +0800

    Refs #42835, Update unixTime function to handle seconds and range parameters properly.

diff --git a/CRM/Contribute/BAO/Widget.php b/CRM/Contribute/BAO/Widget.php
index 775ab97ff..2368128c8 100644
--- a/CRM/Contribute/BAO/Widget.php
+++ b/CRM/Contribute/BAO/Widget.php
@@ -127,7 +127,7 @@ class CRM_Contribute_BAO_Widget extends CRM_Contribute_DAO_Widget {
         }
 
         if ($page['end_date']) {
-          $endDate = CRM_Utils_Date::unixTime($page['end_date']);
+          $endDate = CRM_Utils_Date::unixTime($page['end_date'], true);
           if ($endDate && $endDate < $now) {
             $data['is_active'] = FALSE;
           }
diff --git a/CRM/Contribute/Form/ContributionBase.php b/CRM/Contribute/Form/ContributionBase.php
index 5e6edc573..a9507e52d 100644
--- a/CRM/Contribute/Form/ContributionBase.php
+++ b/CRM/Contribute/Form/ContributionBase.php
@@ -491,7 +491,7 @@ class CRM_Contribute_Form_ContributionBase extends CRM_Core_Form {
 
       //start and end date of the contribution page
       $startDate = CRM_Utils_Date::unixTime(CRM_Utils_Array::value('start_date', $this->_values));
-      $endDate = CRM_Utils_Date::unixTime(CRM_Utils_Array::value('end_date', $this->_values));
+      $endDate = CRM_Utils_Date::unixTime(CRM_Utils_Array::value('end_date', $this->_values), true);
       $now = time();
 
       if ($pcpInfo['contribution_page_id'] != $this->_values['id']) {
diff --git a/CRM/Event/BAO/Event.php b/CRM/Event/BAO/Event.php
index 8ece52ab8..1ea31058c 100644
--- a/CRM/Event/BAO/Event.php
+++ b/CRM/Event/BAO/Event.php
@@ -1809,8 +1809,8 @@ WHERE  ce.loc_block_id = $locBlockId";
 
   static function validRegistrationDate(&$values, $contactID) {
     // make sure that we are between  registration start date and registration end date
-    $startDate = strtotime(CRM_Utils_Array::value('registration_start_date', $values));
-    $endDate = strtotime(CRM_Utils_Array::value('registration_end_date', $values));
+    $startDate = CRM_Utils_Date::unixTime(CRM_Utils_Array::value('registration_start_date', $values));
+    $endDate = CRM_Utils_Date::unixTime(CRM_Utils_Array::value('registration_end_date', $values), true);
     $now = time();
     $validDate = TRUE;
     if ($startDate && $startDate >= $now) {
diff --git a/CRM/Event/BAO/Participant.php b/CRM/Event/BAO/Participant.php
index 422bc04fb..4d1dd3d93 100644
--- a/CRM/Event/BAO/Participant.php
+++ b/CRM/Event/BAO/Participant.php
@@ -1527,7 +1527,7 @@ UPDATE  civicrm_participant
         require_once 'CRM/Contact/BAO/Contact/Utils.php';
         $checksumLife = 'inf';
         if ($endDate = CRM_Utils_Array::value('end_date', $eventDetails)) {
-          $checksumLife = (CRM_Utils_Date::unixTime($endDate) - time()) / (60 * 60);
+          $checksumLife = (CRM_Utils_Date::unixTime($endDate, true) - time()) / (60 * 60);
         }
         $checksumValue = CRM_Contact_BAO_Contact_Utils::generateChecksum($contactId, NULL, $checksumLife);
       }
diff --git a/CRM/Event/Form/ParticipantView.php b/CRM/Event/Form/ParticipantView.php
index 18e77210b..111623ac5 100644
--- a/CRM/Event/Form/ParticipantView.php
+++ b/CRM/Event/Form/ParticipantView.php
@@ -83,7 +83,7 @@ class CRM_Event_Form_ParticipantView extends CRM_Core_Form {
       $checksumLife = 'inf';
       $endDate = CRM_Utils_Array::value('end_date', $eventDetails);
       if ($endDate) {
-        $checksumLife = (CRM_Utils_Date::unixTime($endDate) - time()) / (60 * 60);
+        $checksumLife = (CRM_Utils_Date::unixTime($endDate, true) - time()) / (60 * 60);
       }
       $checksumValue = CRM_Contact_BAO_Contact_Utils::generateChecksum($contactID, NULL, $checksumLife);
       $values[$participantID]['checksum'] = $checksumValue;
diff --git a/CRM/Event/Form/Registration.php b/CRM/Event/Form/Registration.php
index 0366afeae..79891dc16 100644
--- a/CRM/Event/Form/Registration.php
+++ b/CRM/Event/Form/Registration.php
@@ -1093,7 +1093,7 @@ class CRM_Event_Form_Registration extends CRM_Core_Form {
         if(!$this->_values['event']['is_monetary'] && $this->_values['event']['allow_cancel_by_link']){
           $checksumLife = 'inf';
           if ($endDate = CRM_Utils_Array::value('end_date', $this->_values['event'])) {
-            $checksumLife = (CRM_Utils_Date::unixTime($endDate) - time()) / (60 * 60);
+            $checksumLife = (CRM_Utils_Date::unixTime($endDate, true) - time()) / (60 * 60);
           }
           $checksumValue = CRM_Contact_BAO_Contact_Utils::generateChecksum($contactId, NULL, $checksumLife);
           $this->assign('hasCancelLink', TRUE);
diff --git a/CRM/Utils/Date.php b/CRM/Utils/Date.php
index c33d23f92..dbe572f35 100644
--- a/CRM/Utils/Date.php
+++ b/CRM/Utils/Date.php
@@ -248,14 +248,32 @@ class CRM_Utils_Date {
     return $fullMonthNames;
   }
 
-  static function unixTime($string) {
+  /**
+   * Convert date string to Unix timestamp
+   *
+   * If seconds exist in input: Uses original seconds
+   * If no seconds specified: Uses 0 by default, or 59 if $useEndOfMinute=true
+   *
+   * @param string $string Date string (YYYY-MM-DD HH:MM:SS)
+   * @param bool $useEndOfMinute If true, sets seconds to 59 when not specified
+   * @return int Unix timestamp
+   *
+   */
+  static function unixTime($string, $useEndOfMinute = false) {
     if (empty($string)) {
       return 0;
     }
     $parsedDate = date_parse($string);
+
+    if (isset($parsedDate['second']) && $parsedDate['second'] > 0) {
+      $second = CRM_Utils_Array::value('second', $parsedDate, 0);
+    } else {
+      $second = $useEndOfMinute ? 59 : 0;
+    }
+
     return mktime(CRM_Utils_Array::value('hour', $parsedDate),
       CRM_Utils_Array::value('minute', $parsedDate),
-      59,
+      $second,
       CRM_Utils_Array::value('month', $parsedDate),
       CRM_Utils_Array::value('day', $parsedDate),
       CRM_Utils_Array::value('year', $parsedDate)
diff --git a/CRM/Widget/Widget.php b/CRM/Widget/Widget.php
index 53f089313..a0e0992d6 100644
--- a/CRM/Widget/Widget.php
+++ b/CRM/Widget/Widget.php
@@ -139,7 +139,7 @@ WHERE  id = %1";
       }
 
       if ($dao->end_date) {
-        $endDate = CRM_Utils_Date::unixTime($dao->end_date);
+        $endDate = CRM_Utils_Date::unixTime($dao->end_date, true);
         if ($endDate &&
           $endDate < $now
         ) {
